checker/src/org/checkerframework/checker/upperbound/UpperBoundStore.java

```diff
@@ -0,0 +1,157 @@
+            }
+            for (FieldAccess rec : fieldValues.keySet()) {
+                applyAdd(rec, replace, atypeFactory, caller);
+            }
+        }
+        for (Receiver rec : replace.keySet()) {
+            replaceValue(rec, replace.get(rec));
+        }
+
+        super.updateForMethodCall(n, atypeFactory, val);
+    }
+
+    private void applyTransfer(
+            Receiver rec,
+            Map<Receiver, CFValue> replace,
+            boolean isClear,
+            AnnotatedTypeFactory atypeFactory,
+            Receiver caller) {
+
+        UpperBoundAnnotatedTypeFactory factory = (UpperBoundAnnotatedTypeFactory) atypeFactory;
+        CFValue value = this.getValue(rec);
+        Set<AnnotationMirror> atm = value.getAnnotations();
+        if (isClear) {
+            CFValue val = analysis.createSingleAnnotationValue(factory.UNKNOWN, rec.getType());
+            replace.put(rec, val);
+        } else if (AnnotationUtils.containsSameByClass(atm, LTOMLengthOf.class)) {
+            AnnotationMirror anno = AnnotationUtils.getAnnotationByClass(atm, LTOMLengthOf.class);
+            AnnotationMirror newAnno =
+                    factory.createLTLengthOfAnnotation(UpperBoundUtils.getValue(anno));
+            CFValue val =
+                    analysis.createSingleAnnotationValue(
+                            factory.getQualifierHierarchy().leastUpperBound(newAnno, anno),
+                            rec.getType());
+            replace.put(rec, val);
+        } else if (AnnotationUtils.containsSameByClass(atm, LTLengthOf.class)) {
+            AnnotationMirror anno = AnnotationUtils.getAnnotationByClass(atm, LTLengthOf.class);
+            AnnotationMirror newAnno =
+                    factory.createLTEqLengthOfAnnotation(UpperBoundUtils.getValue(anno));
+            CFValue val =
+                    analysis.createSingleAnnotationValue(
+                            factory.getQualifierHierarchy().leastUpperBound(newAnno, anno),
+                            rec.getType());
+            replace.put(rec, val);
+        } else if (AnnotationUtils.containsSameByClass(atm, LTEqLengthOf.class)) {
+            CFValue val = analysis.createSingleAnnotationValue(factory.UNKNOWN, rec.getType());
+            replace.put(rec, val);
+        }
+    }
+
+    private void applyAdd(
+            Receiver rec,
+            Map<Receiver, CFValue> replace,
+            AnnotatedTypeFactory atypeFactory,
+            Receiver caller) {
+
+        UpperBoundAnnotatedTypeFactory factory = (UpperBoundAnnotatedTypeFactory) atypeFactory;
+        CFValue value = this.getValue(rec);
+        Set<AnnotationMirror> atm = value.getAnnotations();
+        if (AnnotationUtils.containsSameByClass(atm, LTEqLengthOf.class)) {
+            AnnotationMirror anno = AnnotationUtils.getAnnotationByClass(atm, LTEqLengthOf.class);
+            String[] vals = UpperBoundUtils.getValue(anno);
+            if (vals.length != 1 || !vals[0].equals(caller.toString())) {
```

Comment:
`vals[0].equals(caller.toString())` is important to the correctness of this method.  Please move it to an if block surrounding the code that is executed if true.  In it's current location it was diffacult for me to understand what was going on.